Atomically write golden file updates

howardjohn opened the pull request:
We have some tests running in parallel writing to the same file, which
leads to some occasional test failures. Ran this 1k times locally, it
passes 100% with my change and almost 0% without on a test of 100 tests
writing to one golden file

Example failure: https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/26774/gencheck_istio/17280

howardjohn committed:
Atomically write golden file updates

We have some tests running in parallel writing to the same file, which
leads to some occasional test failures. Ran this 1k times locally, it
passes 100% with my change and almost 0% without on a test of 100 tests
writing to one golden file

istio-testing commented:
@howardjohn: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-notes_istio | 2e280629bd60d9f0ba4485f1f6d388573410ed9e | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/26802/release-notes_istio/2354) | `/test release-notes_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@howardjohn: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-notes_istio | 2e280629bd60d9f0ba4485f1f6d388573410ed9e | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/26802/release-notes_istio/2354) | `/test release-notes_istio`
lint_istio | 2e280629bd60d9f0ba4485f1f6d388573410ed9e | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/26802/lint_istio/18726) | `/test lint_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@howardjohn: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-notes_istio | 2e280629bd60d9f0ba4485f1f6d388573410ed9e | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/26802/release-notes_istio/2354) | `/test release-notes_istio`
lint_istio | 2e280629bd60d9f0ba4485f1f6d388573410ed9e | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/26802/lint_istio/18726) | `/test lint_istio`
gencheck_istio | 2e280629bd60d9f0ba4485f1f6d388573410ed9e | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/26802/gencheck_istio/17309) | `/test gencheck_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

howardjohn committed:
fixes

rlenglet commented on pilot/test/util/diff.go:
```diff
@@ -110,12 +113,39 @@ func StripVersion(yaml []byte) []byte {
 }
 
+// Write atomically by writing to a temporary file in the same directory then renaming
+func AtomicWrite(path string, data []byte, mode os.FileMode) (err error) {
```
Reuse the existing `AtomicWrite` function used in Istio CNI. It's better, as it doesn't leave the tmp file behind in case of error, etc.
https://github.com/istio/istio/blob/master/cni/pkg/install-cni/pkg/util/util.go#L45

rlenglet replied:
Or was it intentional to leave the tmp file for debugging?

howardjohn committed:
Abstract

howardjohn committed:
Drop dependency on etcd

howardjohn commented:
@rlenglet I moved the CNI stuff out to a common package and used that one. In the process, I found we were importing all of etcd just for ~5 lines of code, which I replicated and dropped that depency

istio-testing commented:
@howardjohn: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio | dbff6112a7b9d803d3dfab5988a7352e08be8a68 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/26802/lint_istio/18735) | `/test lint_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

howardjohn committed:
banner

istio-testing commented:
@howardjohn: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio | dbff6112a7b9d803d3dfab5988a7352e08be8a68 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/26802/lint_istio/18735) | `/test lint_istio`
integ-pilot-k8s-tests_istio | dbff6112a7b9d803d3dfab5988a7352e08be8a68 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/26802/integ-pilot-k8s-tests_istio/18977) | `/test integ-pilot-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@howardjohn: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-pilot-k8s-tests_istio | fd0f7305bb04d42f96248789b3c2d219522144ea | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/26802/integ-pilot-k8s-tests_istio/18980) | `/test integ-pilot-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

howardjohn commented:
/retest

rlenglet commented on tools/istio-iptables/pkg/validation/linux_test.go:
```diff
@@ -17,14 +17,27 @@ package validation
+var byteOrder binary.ByteOrder
+
+// Inspired by etcd cpuutil
+func init() {
```
Remove this. It's a duplicate of what's done in `linux.go` in the same directory.

howardjohn replied:
thanks, removed this

rlenglet commented on tools/istio-iptables/pkg/validation/linux_test.go:
```diff
@@ -17,14 +17,27 @@ package validation
 	hostValue := ntohs(0xbeef)
 	expectValue := 0xbeef
-	if cpuutil.ByteOrder().String() == binary.LittleEndian.String() {
+	if byteOrder == binary.LittleEndian {
```
Just use the `nativeByteOrder` variable already defined in `linux.go`.

rlenglet reviewed (changes requested):
Really nice, thanks! We can dedupe the code a bit more.

howardjohn committed:
dedupe
